xmlgen: Bad code generated for org.gtk.Actions.Changed signal

Seems to be a return of #149, or something similar.

Using the files from #411, I am getting:

```
   --> src/cog.rs:24:1
    |
24  | #[dbus_proxy(interface = "org.gtk.Actions", assume_defaults = true)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `&[&str]`
```

and

```
error[E0277]: the trait bound `&[zbus::zvariant::Value<'_>]: Deserialize<'_>` is not satisfied
   --> src/cog.rs:24:1
    |
24  | #[dbus_proxy(interface = "org.gtk.Actions", assume_defaults = true)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `&[zbus::zvariant::Value<'_>]`
```

Relevant XML excerpt:

```xml
    <signal name="Changed">
      <arg type="as" name="removals">
      </arg>
      <arg type="a{sb}" name="enable_changes">
      </arg>
      <arg type="a{sv}" name="state_changes">
      </arg>
      <arg type="a{s(bgav)}" name="additions">
      </arg>
    </signal>
```

generated function:

```rust
    /// Changed signal
    #[dbus_proxy(signal)]
    fn changed(
        &self,
        removals: &[&str],
        enable_changes: std::collections::HashMap<&str, bool>,
        state_changes: std::collections::HashMap<&str, zbus::zvariant::Value<'_>>,
        additions: std::collections::HashMap<
            &str,
            (bool, zbus::zvariant::Signature<'_>, &[zbus::zvariant::Value<'_>]),
        >,
    ) -> zbus::Result<()>;
```

The same problem generating the `net.Connman.Manager`:

```xml
		<signal name="ServicesChanged">
			<arg type="a(oa{sv})" name="changedServiceList"/>
			<arg type="ao" name="deletedService"/>
		</signal>
```

to:

```rust
    /// ServicesChanged signal
    #[dbus_proxy(signal)]
    fn services_changed(
        &self,
        changedServiceList: &[(
            zbus::zvariant::ObjectPath<'_>,
            std::collections::HashMap<&str, zbus::zvariant::Value<'_>>,
        )],
        deletedService: &[zbus::zvariant::ObjectPath<'_>],
    ) -> zbus::Result<()>;
```

with error:

```
error[E0277]: the trait bound `&[(zbus::zvariant::ObjectPath<'_>, HashMap<&str, zbus::zvariant::Value<'_>>)]: Deserialize<'_>` is not satisfied
   --> src/manager.rs:15:1
    |
15  | #[dbus_proxy(interface = "net.connman.Manager", assume_defaults = true)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `&[(zbus::zvariant::ObjectPath<'_>, HashMap<&str, zbus::zvariant::Value<'_>>)]`
    |
    = help: the following other types implement trait `Deserialize<'de>`:
              [T; 0]
              [T; 1]
              [T; 2]
              [T; 3]
              [T; 4]
              [T; 5]
              [T; 6]
              [T; 7]
            and 26 others
    = note: required for `(&[(zbus::zvariant::ObjectPath<'_>, HashMap<&str, zbus::zvariant::Value<'_>>)], &[zbus::zvariant::ObjectPath<'_>])` to implement `Deserialize<'_>`
    = note: required for `(&[(zbus::zvariant::ObjectPath<'_>, HashMap<&str, zbus::zvariant::Value<'_>>)], &[zbus::zvariant::ObjectPath<'_>])` to implement `DynamicDeserialize<'_>`
note: required by a bound in `zbus::Message::body`
   --> /home/rcc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zbus-3.14.1/src/message.rs:395:12
    |
393 |     pub fn body<'d, 'm: 'd, B>(&'m self) -> Result<B>
    |            ---- required by a bound in this associated function
394 |     where
395 |         B: zvariant::DynamicDeserialize<'d>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Message::body`
    = note: this error originates in the attribute macro `dbus_proxy` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `&[zbus::zvariant::ObjectPath<'_>]: Deserialize<'_>` is not satisfied
   --> src/manager.rs:15:1
    |
15  | #[dbus_proxy(interface = "net.connman.Manager", assume_defaults = true)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `&[zbus::zvariant::ObjectPath<'_>]`
    |
    = help: the following other types implement trait `Deserialize<'de>`:
              [T; 0]
              [T; 1]
              [T; 2]
              [T; 3]
              [T; 4]
              [T; 5]
              [T; 6]
              [T; 7]
            and 26 others
    = note: required for `(&[(zbus::zvariant::ObjectPath<'_>, HashMap<&str, zbus::zvariant::Value<'_>>)], &[zbus::zvariant::ObjectPath<'_>])` to implement `Deserialize<'_>`
    = note: required for `(&[(zbus::zvariant::ObjectPath<'_>, HashMap<&str, zbus::zvariant::Value<'_>>)], &[zbus::zvariant::ObjectPath<'_>])` to implement `DynamicDeserialize<'_>`
note: required by a bound in `zbus::Message::body`
   --> /home/rcc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zbus-3.14.1/src/message.rs:395:12
    |
393 |     pub fn body<'d, 'm: 'd, B>(&'m self) -> Result<B>
    |            ---- required by a bound in this associated function
394 |     where
395 |         B: zvariant::DynamicDeserialize<'d>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Message::body`
    = note: this error originates in the attribute macro `dbus_proxy` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `rconnman_zbus` (bin "rconnman_zbus") due to 2 previous errors
```

I'm seeing this in zbus 4.1.2:
```
error[E0277]: the trait bound `&[&str]: Deserialize<'_>` is not satisfied
  --> src/eds.rs:22:1
   |
22 | / #[proxy(
23 | |     interface = "org.gnome.evolution.dataserver.Calendar",
24 | |     default_service = "org.gnome.evolution.dataserver.Calendar8",
25 | |     gen_blocking = false,
26 | |     assume_defaults = true
27 | | )]
   | |__^ the trait `Deserialize<'_>` is not implemented for `&[&str]`
   |
   = help: the following other types implement trait `Deserialize<'de>`:
             [T; 0]
             [T; 1]
             [T; 2]
             [T; 3]
             [T; 4]
             [T; 5]
             [T; 6]
             [T; 7]
           and 26 others
   = note: required for `&[&str]` to implement `DynamicDeserialize<'_>`
note: required by a bound in `Body::deserialize`
```

--- a/zbus_xmlgen/src/lib.rs
+++ b/zbus_xmlgen/src/lib.rs
@@ -379,14 +379,14 @@ fn to_rust_type(ty: &CompleteType, input: bool, as_ref: bool) -> String {
                 match **c as char {
                     '{' => format!(
                         "std::collections::HashMap<{}>",
-                        iter_to_rust_type(it, input, false)
+                        iter_to_rust_type(it, input, as_ref)
                     ),
                     _ => {
-                        let ty = iter_to_rust_type(it, input, false);
-                        if input {
+                        let ty = iter_to_rust_type(it, input, as_ref);
+                        if input && as_ref {
                             format!("&[{ty}]")
                         } else {
-                            format!("{}Vec<{ty}>", if as_ref { "&" } else { "" })
+                            format!("Vec<{ty}>")
                         }
                     }
                 }
@@ -402,7 +402,7 @@ fn to_rust_type(ty: &CompleteType, input: bool, as_ref: bool) -> String {
                             it.next().unwrap();
                             break;
                         }
-                        _ => vec.push(iter_to_rust_type(it, input, false)),
+                        _ => vec.push(iter_to_rust_type(it, input, as_ref)),
                     }
                 }
                 if dict {

--- a/zbus_xmlgen/src/lib.rs
+++ b/zbus_xmlgen/src/lib.rs
@@ -386,7 +386,7 @@ fn to_rust_type(ty: &CompleteType, input: bool, as_ref: bool) -> String {
                         if input {
                             format!("&[{ty}]")
                         } else {
-                            format!("{}Vec<{}>", if as_ref { "&" } else { "" }, ty)
+                            format!("{}Vec<{ty}>", if as_ref { "&" } else { "" })
                         }
                     }
                 }

--- a/zbus_xmlgen/tests/data/sample_object0.rs
+++ b/zbus_xmlgen/tests/data/sample_object0.rs
@@ -30,7 +30,7 @@ trait SampleInterface0 {
     ) -> zbus::Result<(String, std::collections::HashMap<u32, String>)>;
 
     /// MogrifyMe method
-    fn mogrify_me(&self, bar: &(i32, i32, &[zbus::zvariant::Value<'_>])) -> zbus::Result<()>;
+    fn mogrify_me(&self, bar: &(i32, i32, &[&zbus::zvariant::Value<'_>])) -> zbus::Result<()>;
 
     /// Odyssey method
     #[allow(clippy::too_many_arguments)]

--- a/zbus_xmlgen/tests/data/sample_object0.rs
+++ b/zbus_xmlgen/tests/data/sample_object0.rs
@@ -53,6 +53,21 @@ trait SampleInterface0 {
     #[zbus(signal)]
     fn changed2(&self, new_value: bool, new_value2: bool) -> zbus::Result<()>;
 
+    /// SignalArrayOfStrings signal
+    #[zbus(signal)]
+    fn signal_array_of_strings(&self, array: Vec<&str>) -> zbus::Result<()>;
+
+    /// SignalDictStringToValue signal
+    #[zbus(signal)]
+    fn signal_dict_string_to_value(
+        &self,
+        dict: std::collections::HashMap<&str, zbus::zvariant::Value<'_>>,
+    ) -> zbus::Result<()>;
+
+    /// SignalValue signal
+    #[zbus(signal)]
+    fn signal_value(&self, value: zbus::zvariant::Value<'_>) -> zbus::Result<()>;
+
     /// Bar property
     #[zbus(property)]
     fn bar(&self) -> zbus::Result<u8>;

--- a/zbus_xmlgen/tests/data/sample_object0.xml
+++ b/zbus_xmlgen/tests/data/sample_object0.xml
@@ -43,6 +43,15 @@
        <arg name="new_value" type="b" direction="out"/>
        <arg name="new_value2" type="b" direction="out"/>
      </signal>
+     <signal name="SignalArrayOfStrings">
+       <arg type="as" name="array"/>
+     </signal>
+     <signal name="SignalValue">
+       <arg type="v" name="value"/>
+     </signal>
+     <signal name="SignalDictStringToValue">
+       <arg type="a{sv}" name="dict"/>
+     </signal>
      <property name="Bar" type="y" access="readwrite"/>
      <property name="Foo-Bar" type="y" access="readwrite"/>
      <property name="Matryoshkas" type="a(oiasta{sv})" access="read"/>

